### PR TITLE
Prevent WhitelistingTextInputFormatter to return a empty string if the current value does not satisfy the formatter

### PR DESCRIPTION
**Pull request by pedromassango**
## Description

This PR fixes a bug that actualy exist in the `WhitelistingTextInputFormatter` class: it currently return a empty string if the current value does not satisfy the provided pattern. 

## Related Issues

https://github.com/flutter/flutter/issues/54417

## Tests

I added the following tests: if the new value provided in the formatter does not satisfy the pattern of the formatter then it should return the old value as it is the latest valid value typed by the user.

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

**Comment by pedromassango** [comment]
@goderbauer  Could you help me in review this PR? I tagged you because you have made changes in the file that I'm working on.

File: packages/flutter/lib/src/services/text_formatter.dart [hunk 1]
```diff
@@ -268,6 +271,15 @@ TextEditingValue _selectionAwareTextManipulation(
...
   if (selectionStartIndex < 0 || selectionEndIndex < 0) {
     manipulatedText = substringManipulation(value.text);
+  } else if (value.selection.isCollapsed) { // non-selection text manipulation
+    manipulatedText = substringManipulation(value.text);
+    // We only return the old valid value if the current value is not empty and
+    // if manipulation fails.
+    if (value.text.isNotEmpty && manipulatedText.isEmpty) {
+      manipulatedText = oldValue.text;
+    }
+    // move the carret to the end of the text
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_formatter.dart:281]: "Move the caret to the end of the text."

File: packages/flutter/lib/src/services/text_formatter.dart [hunk 2]
```diff
@@ -268,6 +271,15 @@ TextEditingValue _selectionAwareTextManipulation(
...
+    manipulatedSelection = TextSelection.collapsed(offset: manipulatedText.length);
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_formatter.dart:282]: There is a bug that I believe is caused by this line:

  1. Type a bunch of numbers into the input.
  1. Move the cursor to the middle of the text.
  1. Insert another number at the new cursor position.
  1. The cursor jumps to the end of the line, but it should stay at its position.

**Review by justinmc — COMMENTED** [review]
Thanks for the PR!  This definitely looks like something we should fix.

See the bug that I mentioned.  Could you also include a test that covers that case?

Let me know when you've fixed that and I can take another look.

File: packages/flutter/lib/src/services/text_formatter.dart (hunk 2 above)
Comment (pedromassango) [reply packages/flutter/lib/src/services/text_formatter.dart:282]: Noted. I must work on that soon

Comment (pedromassango) [reply packages/flutter/lib/src/services/text_formatter.dart:282]: Hey @justinmc I was unable to reproduce this issue, can you provide the code that you've used?

Comment (pedromassango) [reply packages/flutter/lib/src/services/text_formatter.dart:282]: I found it :)

Comment (pedromassango) [reply packages/flutter/lib/src/services/text_formatter.dart]: @justinmc I did some updates!

**Comment by pedromassango** [comment]
Hi @justinmc, I'm still wating for your review on this :)

**Comment by justinmc** [comment]
Thanks for the reminder, sorry about that!  Will review this first thing tomorrow morning.

File: packages/flutter/lib/src/services/text_formatter.dart [hunk 3]
```diff
@@ -268,6 +271,19 @@ TextEditingValue _selectionAwareTextManipulation(
...
   if (selectionStartIndex < 0 || selectionEndIndex < 0) {
     manipulatedText = substringManipulation(value.text);
+  } else if (value.selection.isCollapsed) { // non-selection text manipulation
+    int cursorPosition = value.selection.baseOffset;
+    manipulatedText = substringManipulation(value.text);
+    // We only return the old valid value if the current value is not empty and
+    // if manipulation fails.
+    if (value.text.isNotEmpty && manipulatedText.isEmpty) {
+      manipulatedText = oldValue.text;
+      // we decrease cursorPosition by one because
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_formatter.dart:281]: Capital letter "We" and period at the end of the sentence.

File: packages/flutter/lib/src/services/text_formatter.dart [hunk 4]
```diff
@@ -268,6 +271,19 @@ TextEditingValue _selectionAwareTextManipulation(
   TextSelection manipulatedSelection;
   if (selectionStartIndex < 0 || selectionEndIndex < 0) {
     manipulatedText = substringManipulation(value.text);
+  } else if (value.selection.isCollapsed) { // non-selection text manipulation
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_formatter.dart:274]: Capital letter "Non"

**Review by justinmc — COMMENTED** [review]
Could you make these two quick fixes in the comments?  Just some capitalization stuff.

Otherwise this looks great.  Thanks for writing the test for the cursor position.

File: packages/flutter/lib/src/services/text_formatter.dart (hunk 4 above)
Comment (pedromassango) [reply packages/flutter/lib/src/services/text_formatter.dart]: Done :)

**Review by justinmc — APPROVED** [review]
LGTM

I'll try to get this merged when it's green.  Thanks again for the PR!

**Comment by pedromassango** [comment]
> LGTM
> 
> I'll try to get this merged when it's green. Thanks again for the PR!

You welcome. It will help me and others

**Comment by justinmc** [comment]
@pedromassango Can you merge master into this?  I see a failure that I thought was fixed.

**Comment by pedromassango** [comment]
> @pedromassango Can you merge master into this? I see a failure that I thought was fixed.

@justinmc should be fine now

**Comment by justinmc** [comment]
Thanks!  Will merge when the tree is green.

**Pull request merged**